scrape/scrape.go

```diff
@@ -1560,18 +1563,36 @@ loop:
 		// Increment added even if there's an error so we correctly report the
 		// number of samples remaining after relabeling.
 		added++
-
+		exemplarQueue = exemplarQueue[:0] // reset and reuse the exemplar queue
 		for hasExemplar := p.Exemplar(&e); hasExemplar; hasExemplar = p.Exemplar(&e) {
 			if !e.HasTs {
 				e.Ts = t
 			}
+			exemplarQueue = append(exemplarQueue, e)
+			e = exemplar.Exemplar{} // reset for next time round loop
+		}
+		sort.Slice(exemplarQueue, func(i, j int) bool {
+			return exemplarQueue[i].Ts < exemplarQueue[j].Ts
+		})
+		outOfOrderExemplars := 0
+		for _, e := range exemplarQueue {
 			_, exemplarErr := app.AppendExemplar(ref, lset, e)
-			exemplarErr = sl.checkAddExemplarError(exemplarErr, e, &appErrs)
-			if exemplarErr != nil {
+			switch exemplarErr {
+			case nil:
+				// do nothing
+			case storage.ErrOutOfOrderExemplar:
+				outOfOrderExemplars++
+			default:
 				// Since exemplar storage is still experimental, we don't fail the scrape on ingestion errors.
 				level.Debug(sl.l).Log("msg", "Error while adding exemplar in AddExemplar", "exemplar", fmt.Sprintf("%+v", e), "err", exemplarErr)
 			}
-			e = exemplar.Exemplar{} // reset for next time round loop
+		}
+		if outOfOrderExemplars > 0 && outOfOrderExemplars == len(exemplarQueue) {
+			// Only report out of order exemplars if all are out of order, otherwise this was a partial update
+			// to some existing set of exemplars.
```

Comment:
How would that be counted in the old world with classic histograms?

Reply:
For classic histograms, the different buckets are separated into different series, so there's only 1 exemplar per series/bucket for each scrape. So if that 1 is out of order, it's treated as all are out of order, and reported and handled accordingly. But this is less likely unless it's a real OOO case, as the normal repeat exemplars would be treated as duplicates instead (since there's only 1 per series/bucket and scrape, comparing with the latest for that would be enough to detect the duplicate).